internal/hashutil/hashutil.go

```diff
@@ -27,3 +31,176 @@ func Hash(v any) (uint64, error) {
 	}
 	return hashstructure.Hash(v, hashstructure.FormatV2, opts)
 }
+
+type Digest struct {
+	xxhash.Digest
+}
+
+func NewDigest() *Digest {
+	var d Digest
+	d.Reset()
+	return &d
+}
+
+// WriteStringWithLen writes the string's length, then its contents to the hash.
+func (d *Digest) WriteStringWithLen(s string) {
+	d.WriteInt32(int32(len(s)))
+	d.WriteString(s)
+}
+
+// WriteStringWithLen writes the byte array's length, then its contents to
+// the hash.
+func (d *Digest) WriteWithLen(b []byte) {
+	d.WriteInt32(int32(len(b)))
+	d.Write(b)
+}
+
+// WriteBool writes a single byte (1 or 0) to the hash.
+func (d *Digest) WriteBool(b bool) {
+	if b {
+		d.Write([]byte{1})
+	} else {
+		d.Write([]byte{0})
+	}
+}
+
+// WriteUint32 writes a uint16 to the hash.
+func (d *Digest) WriteUint16(t uint16) {
+	var buf [2]byte
+	binary.LittleEndian.PutUint16(buf[:], t)
```

Comment:
nit: arm is big endian, does it make sense to use `binary.NativeEndian` ?

Reply:
The byte order here is arbitrary, but it should probably be consistent.